src/coreclr/jit/morph.cpp

```diff
@@ -353,7 +353,27 @@ GenTree* Compiler::fgMorphCast(GenTree* tree)
     }
     else if (((tree->gtFlags & GTF_UNSIGNED) == 0) && (srcType == TYP_LONG) && varTypeIsFloating(dstType))
     {
-        return fgMorphCastIntoHelper(tree, CORINFO_HELP_LNG2DBL, oper);
+        oper = fgMorphCastIntoHelper(tree, CORINFO_HELP_LNG2DBL, oper);
+
+        if ((dstType == TYP_FLOAT) && (oper->OperGet() == GT_CALL))
+        {
+            // oper is a helper call to CORINFO_HELP_LNG2DBL
+            // but fgMorphCastIntoHelper sets the return type wrong
+            // leading to problems if we try to CSE this call
+            //
+            oper->gtType = TYP_DOUBLE; // CORINFO_HELP_LNG2DBL actually returns a TYP_DOUBLE
```

Comment:
Should there be an issue tracking fixing this so that `LGN2DBL` properly sets `TYP_DOUBLE`?

Reply:
I am not sure.  Perhaps.